src/main/common/parse.c

```diff
@@ -0,0 +1,405 @@
+                argLen = argEnd - argStart;
+            } else {
+                argLen = strlen(argStart);
+            }
+            if (arg->variants) {
+                const char *varStart = arg->variants;
+                unsigned varIx = 0;
+                while (true) {
+                    const char *varEnd = strchr(varStart, '|');
+                    unsigned varLen = 0;
+                    if (varEnd) {
+                        varLen = varEnd - varStart;
+                    } else {
+                        varLen = strlen(varStart);
+                    }
+                    if (argLen == varLen && strncasecmp(varStart, argStart, varLen) == 0) {
+                        initParsedArgWithInt(parsedArg, varIx);
+                        break;
+                    }
+                    if (!varEnd) {
+                        break;
+                    }
+                    varStart = varEnd + 1;
+                    varIx++;
+                }
+                if (!parsedArg->isParsed) {
+                    return parseArgsResultErr(PARSE_STATUS_INVALID_VARIANT, arg->name);
+                }
+            }
+        }
+            break;
+        case ARG_TYPE_INT:
+        {
+            long int value;
+            parseArgsResult_t intRes = parseInt(argStart, arg->name, &value, sep);
+            if (intRes.status != PARSE_STATUS_OK) {
+                return intRes;
+            }
+            initParsedArgWithInt(parsedArg, value);
+            argEnd = intRes.rest;
+        }
+            break;
+        case ARG_TYPE_INT_IN_RANGE:
+        {
+            long int value;
+            parseArgsResult_t intRes = parseInt(argStart, arg->name, &value, sep);
+            if (intRes.status != PARSE_STATUS_OK) {
+                return intRes;
+            }
+            if (value < arg->range.start || value >= arg->range.end) {
+                return parseArgsResultErrNotInRange(arg->name, arg->range.start, arg->range.end);
+            }
+            initParsedArgWithInt(parsedArg, value);
+            argEnd = intRes.rest;
+        }
+            break;
+        case ARG_TYPE_INT_CONVERT:
+        {
+            long int value;
+            parseArgsResult_t intRes = parseInt(argStart, arg->name, &value, sep);
+            if (intRes.status != PARSE_STATUS_OK) {
+                return intRes;
+            }
+            initParsedArgWithInt(parsedArg, value);
+            argEnd = intRes.rest;
+            convertResult_t convertRes = (*arg->convert)(value);
+            if (!convertRes.isOk) {
+                if (convertRes.errMsg) {
+                    return parseArgsResultErrCustomMsg(arg->name, convertRes.errMsg);
+                }
+                return parseArgsResultErr(PARSE_STATUS_INT_CONVERT_ERR, arg->name);
+            }
+            initParsedArgWithInt(parsedArg, convertRes.value);
+        }
+            break;
+        }
+
+        argStart = argEnd;
+        prevArg = arg;
+    }
+
+    ok:
```

Comment:
Finally someone not afraid to use `goto` 👍 

```suggestion
    args_terminated:
```